src/Storages/ObjectStorageQueue/ObjectStorageQueueSource.cpp

```diff
@@ -74,7 +115,101 @@ bool ObjectStorageQueueSource::FileIterator::isFinished() const
 
 size_t ObjectStorageQueueSource::FileIterator::estimatedKeysCount()
 {
-    throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Method estimateKeysCount is not implemented");
+    /// Copied from StorageObjectStorageSource::estimateKeysCount().
+    if (object_infos.empty() && !is_finished && object_storage_iterator->isValid())
+        return std::numeric_limits<size_t>::max();
+    else
+        return object_infos.size();
+}
+
+ObjectStorageQueueSource::Source::ObjectInfoPtr ObjectStorageQueueSource::FileIterator::next()
+{
+    std::lock_guard lock(next_mutex);
+
+    bool current_batch_processed = object_infos.empty() || index >= object_infos.size();
+    if (is_finished && current_batch_processed)
+        return {};
+
+    if (current_batch_processed)
+    {
+        Source::ObjectInfos new_batch;
+        while (new_batch.empty())
+        {
+            auto result = object_storage_iterator->getCurrentBatchAndScheduleNext();
+            if (!result.has_value())
+            {
+                is_finished = true;
+                return {};
+            }
+
+            new_batch = std::move(result.value());
+            for (auto it = new_batch.begin(); it != new_batch.end();)
+            {
+                if (!recursive && !re2::RE2::FullMatch((*it)->getPath(), *matcher))
+                    it = new_batch.erase(it);
+                else
+                    ++it;
+            }
+
+            if (filter_expr)
+            {
+                std::vector<String> paths;
+                paths.reserve(new_batch.size());
+                for (const auto & object_info : new_batch)
+                    paths.push_back(Source::getUniqueStoragePathIdentifier(*configuration, *object_info, false));
+
+                VirtualColumnUtils::filterByPathOrFile(
+                    new_batch, paths, filter_expr, virtual_columns, getContext());
+
+                LOG_TEST(logger, "Filtered files: {} -> {}", paths.size(), new_batch.size());
+
+                size_t previous_size = new_batch.size();
+                new_batch = filterOutProcessedAndFailed(new_batch);
+
+                LOG_TEST(logger, "Filtered files: {} -> {}", previous_size, new_batch.size());
+            }
+            else
+            {
+                size_t previous_size = new_batch.size();
+                new_batch = filterOutProcessedAndFailed(new_batch);
+                LOG_TEST(logger, "Filtered files: {} -> {}", previous_size, new_batch.size());
+            }
+        }
+
+        index = 0;
+        object_infos = std::move(new_batch);
+    }
+
+    if (index >= object_infos.size())
+    {
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR,
+            "Index out of bound for blob metadata. Index: {}, size: {}",
+            index, object_infos.size());
+    }
+
+    return object_infos[index++];
+}
+
+ObjectStorageQueueSource::Source::ObjectInfos
+ObjectStorageQueueSource::FileIterator::filterOutProcessedAndFailed(Source::ObjectInfos & objects)
```

Comment:
```suggestion
ObjectStorageQueueSource::FileIterator::filterOutProcessedAndFailed(Source::ObjectInfos && objects)
```

so you can `std::move(objects[idx])` into result